lib/Base/NumpyReader.cpp

```diff
@@ -0,0 +1,302 @@
+                                       "should match the number of channels: "
+                                    << numCh;
+  size_t chStride = imageLayout == ImageLayout::NCHW ? shape[2] * shape[3] : 1;
+  for (size_t i = 0; i < data.size(); i++) {
+    size_t chIdx = (i / chStride) % numCh;
+    CHECK(data[i] >= 0. && data[i] <= 255.)
+        << "NPY loader: U8 data expected, got: " << data[i];
+    data[i] = (data[i] - meanVal[chIdx]) / stddevVal[chIdx];
+    data[i] = data[i] * scale + bias;
+  }
+}
+
+void loadUnspecifiedImageAndPreprocess(
+    const llvm::ArrayRef<std::string> &filenames, Tensor &inputData,
+    ImageNormalizationMode imageNormMode, llvm::ArrayRef<float> mean,
+    llvm::ArrayRef<float> stddev) {
+  CHECK_EQ(filenames.size(), 1) << "NPY raw image loader: expect single file.";
+  CHECK_LE(mean.size(), 1) << "NPY raw image loader: expect single mean value.";
+  CHECK_LE(stddev.size(), 1)
+      << "NPY raw image loader: expect single stddev value.";
+
+  NpyData dataNpy;
+  numpyReader(filenames[0], dataNpy);
+  std::vector<float> data(dataNpy.nvals);
+  if (dataNpy.type == NpyType::F4) {
+    memcpy(data.data(), dataNpy.data.data(), dataNpy.nvals * dataNpy.elemSize);
+  } else {
+    convertU8S8ToFloat(dataNpy, data);
+  }
+
+  normalizeData(ImageLayout::Unspecified, mean, stddev, imageNormMode, data,
+                dataNpy.shape);
+
+  inputData.reset(ElemKind::FloatTy, dataNpy.shape);
+  inputData.getHandle<>() = data;
+}
+
+void glow::loadNumpyImagesAndPreprocess(
+    const llvm::ArrayRef<std::string> &filenames, Tensor &inputData,
+    ImageNormalizationMode imageNormMode, ImageLayout imageLayout,
+    ImageLayout inputLayout, llvm::ArrayRef<float> mean,
+    llvm::ArrayRef<float> stddev) {
+  DCHECK(!filenames.empty())
+      << "NPY loader: There must be at least one filename in filenames.";
+
+  if (imageLayout == ImageLayout::Unspecified) {
+    return loadUnspecifiedImageAndPreprocess(filenames, inputData,
+                                             imageNormMode, mean, stddev);
+  }
+
+  dim_t numImg = filenames.size();
+
+  // Read each tensor file into a vector of tensors.
+  std::vector<Tensor> tensors(numImg);
+  dim_t batchSize = 0;
+  for (dim_t n = 0; n < numImg; n++) {
+
+    NpyData dataNpy;
+    numpyReader(filenames[n], dataNpy);
+    std::vector<float> data(dataNpy.nvals);
+    convertU8S8ToFloat(dataNpy, data);
+    // Expand 3D to 4D. Supporting 4D only.
+    if (dataNpy.shape.size() == 3) {
+      dataNpy.shape.insert(dataNpy.shape.begin(), 1);
+    }
+    CHECK_EQ(dataNpy.shape.size(), 4)
+        << "NPY loader: Supporting only 3 or 4 dimensions.";
+    normalizeData(imageLayout, mean, stddev, imageNormMode, data,
+                  dataNpy.shape);
+    // Load tensor from the vector obtained from the npy loader.
+    tensors[n].reset(glow::ElemKind::FloatTy, dataNpy.shape);
+    tensors[n].getHandle<>() = data;
+    auto dims0 = tensors[0].dims();
```

Comment:
nit: move this outside the loop?

Reply:
Didn't understand why - performance wise should be the same as aliasing rules will decide if the compiler will move it out of loop itself. Coding style ? Note that tensor is not yet created there.

Reply:
Ah nevermind, misread it.